src/main/java/org/spongepowered/common/user/SpongeUserManager.java

```diff
@@ -24,96 +24,404 @@
+        this.setupWatchers();
     }
 
     @Override
-    public Optional<User> find(final UUID uniqueId) {
-        return this.serverUserProvider.getUser(uniqueId);
+    public CompletableFuture<Optional<User>> load(final UUID uniqueId) {
+        return this.fetchUser(uniqueId, false)
+                .thenApply(x -> {
+                    if (x != null) {
+                        return Optional.of(new SpongeUserView(uniqueId));
+                    }
+                    return Optional.empty();
+                });
     }
 
     @Override
-    public Optional<User> find(final String lastKnownName) {
-        checkNotNull(lastKnownName, "lastKnownName");
-        checkArgument(!lastKnownName.isEmpty() && lastKnownName.length() <= 16, "Invalid username %s", lastKnownName);
-        return this.serverUserProvider.getUser(lastKnownName);
+    public CompletableFuture<User> loadOrCreate(final UUID uuid) {
+        return this.fetchUser(uuid, true);
     }
 
-    @Override
-    public Optional<User> find(final GameProfile profile) {
-        return this.serverUserProvider.getUser(profile);
+    private CompletableFuture<@Nullable User> fetchUser(final UUID uniqueId, final boolean always) {
+        final UUID uuidToUse = this.ensureNonEmptyUUID(uniqueId);
+        if (this.server.getPlayerList().getPlayer(uniqueId) != null) {
+            return CompletableFuture.completedFuture(new SpongeUserView(uniqueId));
+        }
+        final @Nullable SpongeUserData currentUser = this.userCache.getIfPresent(uuidToUse);
+        if (currentUser != null) {
+            return CompletableFuture.completedFuture(new SpongeUserView(uuidToUse));
+        }
+        return CompletableFuture.supplyAsync(() -> {
+            if (always || this.knownUUIDs.contains(uuidToUse)) {
+                final com.mojang.authlib.@Nullable GameProfile profile = this.server.getProfileCache().get(uuidToUse);
+                try {
+                    this.createUser(profile == null ? new com.mojang.authlib.GameProfile(uuidToUse, null) : profile);
+                } catch (final IOException e) {
+                    throw new CompletionException(e);
+                }
+                return new SpongeUserView(uuidToUse);
+            }
+            return null;
+        }, this.executorService);
     }
 
     @Override
-    public User findOrCreate(final GameProfile profile) {
-        return this.serverUserProvider.getOrCreateUser(this.ensureNonEmptyUUID(profile), false);
+    public CompletableFuture<Optional<User>> load(final String lastKnownName) {
+        Objects.requireNonNull(lastKnownName, "lastKnownName");
+        if (lastKnownName.isEmpty() || lastKnownName.length() <= 16) {
+            throw new IllegalArgumentException(String.format("Invalid username %s", lastKnownName));
+        }
+        final com.mojang.authlib.@Nullable GameProfile mcProfile =
+                this.server.getProfileCache().get(lastKnownName.toLowerCase(Locale.ROOT));
```

Comment:
Since this is  using a `ConcurrentMap` under the hood and should be safe to call off thread, can we wrap this in a CompletableFuture and chain the result ?

Reply:
I actually don't want it to go to Mojang for this so I'll just change the call.

Reply:
Fair enough